Add LDAP groups provider plugin

mosiac1 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

Adds an implementation of the Groups Provider plugin type that loads user groups from LDAP. 
Supports filtering of fetched groups at the LDAP query level. 
Supports fetching of extra users attributes, which are mapped to groups with the pattern `{attribute}={value}`.

The plugin uses a 2-pass approach:
1. Find the user object in LDAP, get CN and extra attributes;
2. Search for groups that the user is member of.

Result are cached with a configurable TTL. The cache has extra logic to save loaded groups only when both steps complete successfully, otherwise returning either empty groups of results from step 1.

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues

Sample configuration:

```properties
group-provider.name=ldap
ldap.url=ldaps://my-ldap-server:636
ldap.admin-user=CN=ro-query,OU=users,DC=trino,DC=io
ldap.admin-password=ro-password
ldap.user-base-dn=OU=users,DC=trino,DC=io
ldap.user-search-filter=(&(objectClass=person)(|(name={0})(mail={0}@trino.io)))
ldap.user-search-attributes=uuid
ldap.groups-base-dn=OU=groups,DC=trino,DC=io
ldap.groups-search-filter=(|(cn=dev*)(cn=admin*))
ldap.groups-search-member-attribute=member
ldap.groups-name-attribute=cn
ldap.cache-ttl=10m
```

The PR also updates the Trino LDAP toolkit to support pass-through search filter arguments to the LDAP client. This lets the LDAP server handle argument substitutions.


<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
(x) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

Praveen2112 commented on lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/ldap/LdapQuery.java:
```diff
@@ -21,14 +21,17 @@ public class LdapQuery
         this.searchFilter = requireNonNull(searchFilter, "searchFilter is null");
         requireNonNull(attributes, "attributes is null");
         this.attributes = Arrays.copyOf(attributes, attributes.length);
+        requireNonNull(filterArguments, "filter arguments is null");
+        this.filterArguments = Arrays.copyOf(filterArguments, filterArguments.length);
```
nit: Can we swap the field order so as to be in sync with the ctor ?

Praveen2112 commented on plugin/trino-ldap-groups-provider/pom.xml:
```diff
@@ -0,0 +1,157 @@
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>trino-ldap-groups-provider</artifactId>
```
nit: Should this be `trino-ldap-group-provider`

Praveen2112 commented on plugin/trino-ldap-groups-provider/pom.xml:
```diff
@@ -0,0 +1,157 @@
+            <artifactId>trino-spi</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
```
A new line break and can we move `<!-- for testing - Trino -->` here

Praveen2112 commented on plugin/trino-ldap-groups-provider/pom.xml:
```diff
@@ -0,0 +1,157 @@
+            <groupId>io.airlift</groupId>
+            <artifactId>units</artifactId>
+        </dependency>
+        <dependency>
```
new line break

Praveen2112 commented on plugin/trino-ldap-groups-provider/src/main/java/io/trino/plugin/ldapgroups/LdapGroupsProvider.java:
```diff
@@ -0,0 +1,149 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class LdapGroupsProvider
```
nit: LdapGroupProvider and at other places

Praveen2112 commented on plugin/trino-ldap-groups-provider/src/main/java/io/trino/plugin/ldapgroups/LdapGroupsProvider.java:
```diff
@@ -0,0 +1,149 @@
+
+        this.ldapHighLevelClient = ldapHighLevelClient;
+
+        this.groupListCache = EvictableCacheBuilder.newBuilder()
```
Instead of having the caching as a part of `LdapGroupProvider` - can we introduce something like a `CachingGroupProvider` in the `plugin-toolkit` which could wrap on this implementation

mosiac1 replied:
Created a `CachingGroupProvider` in `plugin-toolkit` with its won `CacheConfig` and module for injecting

Praveen2112 commented on plugin/trino-ldap-groups-provider/src/main/java/io/trino/plugin/ldapgroups/LdapGroupsProvider.java:
```diff
@@ -0,0 +1,149 @@
+                });
+    }
+
+    private Set<String> getLdapGroups(String user)
```
It would be nice if we could move the `private` methods after all the public methods

Praveen2112 commented on plugin/trino-ldap-groups-provider/src/main/java/io/trino/plugin/ldapgroups/LdapGroupsProvider.java:
```diff
@@ -0,0 +1,149 @@
+    private Set<String> getLdapGroups(String user)
+            throws LdapGroupsLoadingException
+    {
+        ImmutableSet.Builder<String> resBuilder = ImmutableSet.builder();
```
Can we avoid abbrevation and instead have `groupsBuilder`

Praveen2112 commented on plugin/trino-ldap-groups-provider/src/main/java/io/trino/plugin/ldapgroups/LdapGroupsProvider.java:
```diff
@@ -0,0 +1,149 @@
+            }
+            catch (ExecutionException e) {
+                Throwable cause = e.getCause();
+                if (cause instanceof LdapGroupsLoadingException exception) {
```
It would be nice if we could fail explicitly when fetching groups instead of silently returning partial set of groups. WDYT ?

mosiac1 replied:
IIRC group providers always fail silently since Trino will ignore errors raised by them and continue query execution (or what else it was doing). This is why I though returning _some_ values is better than failing outright.

I am not convinced this is the best approach, so if you think it needs to be changed I'm ok with it

mosiac1 replied:
Ended up removing this functionality, there was no clean way of using it in conjunction with the `CachingGroupProvider`

Praveen2112 commented on plugin/trino-ldap-groups-provider/src/main/java/io/trino/plugin/ldapgroups/LdapGroupsProviderConfig.java:
```diff
@@ -0,0 +1,36 @@
+import io.airlift.units.Duration;
+import jakarta.validation.constraints.NotNull;
+
+public class LdapGroupsProviderConfig
```
Can we move it to a dedicated config for Caching ?

Praveen2112 commented on plugin/trino-ldap-groups-provider/src/test/java/io/trino/plugin/ldapgroups/TestLdapGroupsProviderIntegration.java:
```diff
@@ -0,0 +1,270 @@
+    {
+        var config = ImmutableMap.<String, String>builder()
+                .putAll(baseConfig)
+                .put("ldap.user-search-filter", "(uid={0})")
```
Can we make the search filter similar to what we configure in the Authenticator - Like in case of LdapAuthenticator - we use `${USER}` which will replace the username pattern. Can we reuse the same approach here ?

mosiac1 replied:
This is using `{0}` so that the substitution can be done by the LDAP client/server, compared to the LDAP password authentication which handles substitution on its own.

This is also the reason behind the change to [`JdkLdapClient`](https://github.com/trinodb/trino/pull/20157/files#diff-c99031194fd25f675a123d6f9648f402e0b0f4dc3356cf574c31459ed96bd607R181).

I would expect this approach to be more secure, but I can't find much documentation on it.

Praveen2112 commented on plugin/trino-ldap-groups-provider/src/main/java/io/trino/plugin/ldapgroups/LdapHighLevelClient.java:
```diff
@@ -0,0 +1,164 @@
+        this.groupsSearchFilter = config.getLdapGroupsSearchFilter();
+        this.groupsNameAttribute = config.getLdapGroupsNameAttribute();
+        this.userSearchFilter = config.getLdapUserSearchFilter();
+        this.userSearchAttributes = config.getLdapUserSearchAttributes();
```
ImmutableSet#copyOf

Praveen2112 commented on plugin/trino-ldap-groups-provider/src/main/java/io/trino/plugin/ldapgroups/LdapHighLevelClient.java:
```diff
@@ -0,0 +1,164 @@
+     * @return Names of groups that the user is a member of
+     * @throws NamingException Underlying LDAP error
+     */
+    public List<String> getGroups(String userDN)
```
Apart from this approach we also have another one - for AD kind of solution i.e if `memberOf` attribute is present then we could search for the user in a single LDAP query. Though it is out of scope for this PR - It would be helpful if we could proper abstraction so that it could also be benefited. WDYT ?

mosiac1 replied:
We tried using this solution, but ran into two issues:
-  using `memberOf` will fetch ALL groups that a user is a member of. This can be bad for both performance and security reasons;
- `memberOf` attributes _usually_ (depends on the LDAP application) contain the fully qualified name of the group (e.g., `cn=clients,ou=groups,dc=trino,dc=io`)

But I can see there are cases where these are not issues and having a one-query solution would be nice. 

The `LdapGroupsMapping` implementation from the Hadoop libraries allows you to configure a `memberOfAttribute` and will use one query if that is specified - WDYT of doing something similar?

Praveen2112 commented:
@mosiac1 Thank you for addressing the comments !! Will take a look at this PR

mosiac1 commented:
Hello @Praveen2112! I rebased on top of main, split the changes into 3 commits and added tests for the CachingGroupProvider classes.

Would really appreciate if you can take a look!

Praveen2112 commented on plugin/trino-ldap-group-provider/src/main/java/io/trino/plugin/ldapgroup/LdapHighLevelClient.java:
```diff
@@ -0,0 +1,165 @@
+    public LdapHighLevelClient(LdapClient ldapClient, LdapHighLevelClientConfig config)
+    {
+        this.ldapClient = requireNonNull(ldapClient, "ldap client is null");
+        requireNonNull(config, "high level ldap client is null");
```
We can also remove the rnn check for config object

Praveen2112 commented on plugin/trino-ldap-group-provider/src/main/java/io/trino/plugin/ldapgroup/LdapHighLevelClient.java:
```diff
@@ -0,0 +1,165 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class LdapHighLevelClient
```
We can also provide some sort of an abstraction like an interface for `LdapGroupProviderClient` and having multiple implementation one this one and other could be for situations where we have memberOf attribute

Praveen2112 commented on plugin/trino-ldap-group-provider/src/main/java/io/trino/plugin/ldapgroup/LdapHighLevelClient.java:
```diff
@@ -0,0 +1,165 @@
+                            log.warn("User [%s] does not have user search attribute [%s]", user, attributeName);
+                            continue;
+                        }
+                        NamingEnumeration<?> attributeValues = attribute.getAll();
+                        ImmutableList.Builder<String> valuesBuilder = ImmutableList.builder();
+                        while (attributeValues.hasMore()) {
+                            valuesBuilder.add(attributeValues.next().toString());
+                        }
+                        List<String> values = valuesBuilder.build();
```
How about 
```
                        List<String> values = Streams.stream(attribute.getAll().asIterator())
                                .map(Object::toString)
                                .collect(toImmutableList());

Praveen2112 commented on plugin/trino-ldap-group-provider/src/main/java/io/trino/plugin/ldapgroup/LdapHighLevelClient.java:
```diff
@@ -0,0 +1,165 @@
+    public List<String> getGroups(String userDN)
+            throws NamingException
+    {
+        String combinedGroupSearchFilter;
+        if (groupsSearchFilter != null && !groupsSearchFilter.isBlank()) {
+            combinedGroupSearchFilter = String.format("(&(%s)(%s={0}))", groupsSearchFilter, groupsSearchMemberAttribute);
+        }
+        else {
+            combinedGroupSearchFilter = String.format("(%s={0})", groupsSearchMemberAttribute);
+        }
```
Can we move this to the constructor.

Praveen2112 commented on plugin/trino-ldap-group-provider/src/test/java/io/trino/plugin/ldapgroup/TestLdapGroupProvider.java:
```diff
@@ -0,0 +1,116 @@
+import static org.mockito.Mockito.when;
+
+@TestInstance(TestInstance.Lifecycle.PER_METHOD)
+class TestLdapGroupProvider
```
Can we move the mock tests to the integration tests as well - Instead of creating a mock object we could use the images.

Praveen2112 commented on plugin/trino-ldap-group-provider/src/test/java/io/trino/plugin/ldapgroup/TestLdapGroupProviderIntegration.java:
```diff
@@ -0,0 +1,270 @@
+    private final LdapGroupProviderFactory factory = new LdapGroupProviderFactory();
+
+    @Container
+    private static final GenericContainer<?> openldapContainer = new GenericContainer<>("osixia/openldap:1.5.0")
```
Can we use this `TestingOpenLdapServer` - maybe we could move it to toolkit and reuse them for password and group provider.

Praveen2112 reviewed (commented):
Instead of mock library - can we use the TestingOpenLdapServer which could provide a accurate behavior.

ksobolew commented:
Hi @mosiac1,

Could you take a look at #18174 and see if it is useful for you in "outsourcing" the caching to a generic layer? I can see that it's quite similar to what you have done.

mosiac1 commented:
Hey @ksobolew,

https://github.com/trinodb/trino/pull/18174 looks like a superset of what I added with more configuration, so I would be very happy to move to that.

Please ping me once that gets merged and I'll change this PR to use it. I'll keep mine around until then in case you want to take a look at it.

bazooka720 commented:
This would be great to have. Any ETA of this?

kokosing commented:
Please cherry-pick. This is good idea.

mosiac1 commented:
Addressed all the review comments and moved some of the components around. Two big changes:
1. This is now using @ksobolew 's CachingGroupProviderModule;
2. The plugin now support two methods for fetching groups from LDAP:
    *   A 1-pass approach, fetching the `memberOf` attribute which in some AD systems has multiple values representing the groups the user is a member of. The group's name is retrieved from its full DN. Note that this does not support filtering of groups or much customization for what represents the group's name.
    *  The 2-pass approach initially described in the PR description, get the user CN and attributes and search groups with `member` filter.
    
 The choice between the two mechanisms is done by setting the `ldap.user-member-of-attribute` configuration property.
 
 Requested by @Praveen2112 I moved the `TestingOpenLdapContainer` into trino-testing-containers so it could be used by both the password-authenticator and group-provider tests.

mosiac1 commented:
@Praveen2112 Please take a look when you have time. The failing test is part of the Hive suite, I assume its being flaky.

Praveen2112 commented:
One minor nit : I have is can we add the caching logic as a final commit, once the base LDAP implementation has been added ?

Praveen2112 commented on plugin/trino-ldap-group-provider/pom.xml:
```diff
@@ -0,0 +1,158 @@
+    <parent>
+        <groupId>io.trino</groupId>
+        <artifactId>trino-root</artifactId>
+        <version>439-SNAPSHOT</version>
```
Can we update the version ?

Praveen2112 commented on plugin/trino-ldap-group-provider/src/main/java/io/trino/plugin/ldapgroup/LdapGroupProviderConfig.java:
```diff
@@ -0,0 +1,126 @@
+
+import static com.google.common.base.Strings.nullToEmpty;
+
+public class LdapGroupProviderConfig
```
Can we add a few `ConfigDescription`

Praveen2112 commented on plugin/trino-ldap-group-provider/src/main/java/io/trino/plugin/ldapgroup/LdapGroupProviderFactory.java:
```diff
@@ -0,0 +1,72 @@
+        Bootstrap app = new Bootstrap(
+                new LdapClientModule(),
+                CachingGroupProviderModule.create(),
+                binder -> {
+                    configBinder(binder).bindConfig(LdapGroupProviderConfig.class);
+                    binder.bind(GroupProvider.class).annotatedWith(ForCachingGroupProvider.class).to(LdapGroupProvider.class).in(Scopes.SINGLETON);
+                },
+                conditionalModule(
+                        LdapGroupProviderConfig.class,
+                        config -> config.getLdapUserMemberOfAttribute() != null,
+                        binder -> binder.bind(LdapGroupProviderClient.class).to(LdapGroupProviderSimpleClient.class).in(Scopes.SINGLETON)),
+                conditionalModule(
+                        LdapGroupProviderConfig.class,
+                        config -> config.getLdapUserMemberOfAttribute() == null,
+                        binder -> {
+                            configBinder(binder).bindConfig(LdapGroupProviderFilteringClientConfig.class);
+                            binder.bind(LdapGroupProviderClient.class).annotatedWith(ForLdapUserSearchDelegate.class).to(LdapGroupProviderSimpleClient.class).in(Scopes.SINGLETON);
+                            binder.bind(LdapGroupProviderClient.class).to(LdapGroupProviderFilteringClient.class).in(Scopes.SINGLETON);
+                        }));
```
Can we extract them into a seperate module like `LdapGroupProviderModule`

Praveen2112 commented on plugin/trino-ldap-group-provider/src/test/java/io/trino/plugin/ldapgroup/MockLdapClient.java:
```diff
@@ -0,0 +1,71 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class MockLdapClient
```
Can we avoid MockLdapClient ? Instead we could use a LDAP server via test containers to perform additional assertion.

mosiac1 replied:
We have integration tests as well that use an LDAP container, see `plugin/trino-ldap-group-provider/src/test/java/io/trino/plugin/ldapgroup/TestLdapGroupProviderIntegration.java`
Mocking is used to test error handing and behaviour that can't easly be achieved with an actual LDAP server (e.g., malformed data, exceptions)

Praveen2112 replied:
Can we rename it as `TesingLdapClient` ?

mosiac1 commented:
Ok, just to  make sure I get this right, the commits would be (not actual commit message, just for demonstartion):

* Move TestingOpenLdapContainer to trino-testing-containers
* Add LDAP filter argument support to JdkLdapClient 
* Add LDAP group provider
* Add caching group provider in plugin-toolkit
* Update Ldap group provider to use caching

Or should I squash the last two into:

* ...
* Add caching group provider and enable for LDAP group provider

Praveen2112 commented:
It could be like 
```

    Move TestingOpenLdapContainer to trino-testing-containers
    Add LDAP filter argument support to JdkLdapClient
    Add LDAP group provider
    Add caching group provider and enable for LDAP group provider

mosiac1 commented:
Addressed all the comments, please take a look

Praveen2112 commented on plugin/trino-ldap-group-provider/src/test/java/io/trino/plugin/ldapgroup/TestLdapGroupProviderIntegration.java:
```diff
@@ -0,0 +1,314 @@
+        assertEquals(expectedGroups, groups);
+    }
+
+    @ParameterizedTest
```
Can we avoid using `ParametermizedTest`, it makes harder to debug the tests.

mosiac1 replied:
The integrations tests already use `ParametermizedTest` to get a cross-product of configurations to tests (e.g., using `ldap.user-member-of-attribute` or `ldap.group-base-dn`) - I don't see any better way of doing this

Praveen2112 commented on plugin/trino-ldap-group-provider/src/test/java/io/trino/plugin/ldapgroup/MockNamingEnumeration.java:
```diff
@@ -0,0 +1,76 @@
+import java.util.Iterator;
+import java.util.List;
+
+public final class MockNamingEnumeration<T>
```
Can we rename it as `TestingNamingEnumeration`

mosiac1 commented on plugin/trino-ldap-group-provider/pom.xml:
```diff
@@ -0,0 +1,158 @@
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>5.1.1</version>
+            <scope>test</scope>
+        </dependency>
```
This was actually not used, not sure why maven-checks didn't flag it. Removed it

mosiac1 commented on plugin/trino-ldap-group-provider/src/main/java/io/trino/plugin/ldapgroup/LdapGroupProvider.java:
```diff
@@ -0,0 +1,92 @@
+            LdapSearchUserResult userSearchResult = userSearchResultOpt.get();
+            log.debug("Fetched user DN and search attributes from LDAP for user %s - %s", user, userSearchResult);
+
+            groupsBuilder.addAll(userSearchResult
+                    .userAttributes()
+                    .entrySet()
+                    .stream()
+                    .filter(e -> !e.getKey().equalsIgnoreCase(userMemberOfAttribute))
+                    .map(e -> e
+                            .getValue()
+                            .stream()
+                            .map(v -> String.format("%s=%s", e.getKey(), v))
+                            .toList())
+                    .flatMap(List::stream)
+                    .iterator());
```
Yep, that is one the of the main point of this group provider, its mentioned in the pr description

mosiac1 commented on plugin/trino-ldap-group-provider/src/main/java/io/trino/plugin/ldapgroup/LdapGroupProviderConfig.java:
```diff
@@ -0,0 +1,134 @@
+        return this;
+    }
+
+    @NotNull
+    public Set<String> getLdapUserSearchAttributes()
+    {
+        return ldapUserSearchAttributes;
+    }
+
+    @Config("ldap.user-search-attributes")
+    @ConfigDescription("User attributes loaded as groups. Example: cn, uuid")
+    public LdapGroupProviderConfig setLdapUserSearchAttributes(String ldapUserSearchAttributes)
+    {
+        this.ldapUserSearchAttributes = ImmutableSet.copyOf(Splitter.on(",").omitEmptyStrings().trimResults().split(nullToEmpty(ldapUserSearchAttributes)));
+        return this;
+    }
```
please consolidate the other comments related to this one and reword them, the point you are trying to make is very unclear

mosiac1 commented on testing/trino-testing-containers/src/main/java/io/trino/testing/containers/TestingOpenLdapServer.java:
```diff
@@ -76,33 +72,57 @@ public String getNetworkAlias()
+        return createOrganization(name, new DisposableSubContext(BASE_DISTINGUISED_NAME));
+    }
+
+    public DisposableSubContext createOrganization(String name, DisposableSubContext subOrganization)
```
I don't think its worth it to add an extra commit for a new api that isn't going to be used for the next 2 commits. WDYT?

mosiac1 commented on plugin/trino-ldap-group-provider/src/main/java/io/trino/plugin/ldapgroup/LdapGroupProviderFilteringClient.java:
```diff
@@ -0,0 +1,117 @@
+
+        String groupsSearchMemberAttribute = filteringConfig.getLdapGroupsSearchMemberAttribute();
+        String groupsSearchFilter = filteringConfig.getLdapGroupsSearchFilter();
+        if (groupsSearchFilter != null && !groupsSearchFilter.isBlank()) {
```
I tried this and it doesn't work as expected; having `Optional<@NotEmpty String>` as the return type of a getter in the config class will raises errors if `ldap.group-search-filter` is  not set. Keeping `Optional<...>` for now

Praveen2112 commented:
I would like to have a discussion on 

>  Supports fetching of extra users attributes, which are mapped to groups with the pattern {attribute}={value}.

Ideally Group is an LDAP object while these users attributes are a part of the LDAP object. This group can be used for any access controls does it makes sense to rely on a non LDAP object here ? WDYT ?

mosiac1 commented:
Let me know if you want to catch up offline about this, I'm on the Trino slack.

Keep in mind this is configurable. By default no user attributes are mapped to groups. The plugin needs to be configured with what attributes to map. See the tests for examples.

I don't see what the issue is. To Trino (and to most systems for that matter) grops are just strings; groups gain meaning when access control policies are bound to them. Downstream from the group provider it is irelevant where the groups came from.

We already use this at Bloomberg as part of our access control systems so we can create policies more nuanced than "is user _x_ part of group _y_"

Praveen2112 commented on plugin/trino-ldap-group-provider/src/test/java/io/trino/plugin/ldapgroup/TestLdapGroupProviderIntegration.java:
```diff
@@ -0,0 +1,314 @@
+        closer.close();
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideUsersAndExpectedGroups")
+    public void testGetGroups(ConfigBuilder configBuilder, String userName, Set<String> expectedGroups)
```
It is not generally recommended to use `ParameterizedTest` - In general, they make it harder to debug tests - We could try to create some sort of an `assertYYY` - For reference we could try something like `TestDecimalCoercers`

mosiac1 replied:
will look into this a bit, but I personally don't see an overall ban on `ParametrizedTests` as a good thing. they may be overused but they do have a place

Praveen2112 reviewed (commented):
I'm a bit confused on different mechanism being used for user attributes for `ldap.user-search-attributes` we use it key and value while for others it is just `value` - Can it be symmetric ?

mosiac1 commented on plugin/trino-ldap-group-provider/src/main/java/io/trino/plugin/ldapgroup/LdapGroupProviderSimpleClient.java:
```diff
@@ -0,0 +1,160 @@
+            log.warn("Group with DN [%s] has no Relative Domain Names. Ignoring group.", groupDN);
+            return Optional.empty();
+        }
+        Rdn lastRDN = groupRDNs.getLast();
```
an LDAP Distinguished Name (DN) is a chain of multiple Relative Distinguished Names (RDNs). for example in `cn=users,ou=groups,dc=trino,dc=io` - `cn=users` and `ou=groups` are RDNs.

these are meant to be read from right to left (you can notice they get more specific). `getRdns()` will return them in this order, so the last element is the least specific RDN, usually what we would call the "name". in the example above it would be `cn=users`.

parsing group names from DNs is an heuristic based on the assumption that the last RDN is the Canonical Name we are looking for.

mosiac1 commented:
@Praveen2112 in reply to your review comment and  https://github.com/trinodb/trino/pull/20157#discussion_r1578206980_

No, it cannot be symmetric. We prefix attributes from `ldap.user-search-attributes` with the key so we can make sense of the values - what meaning do the values in Map have without their keys? 
We treat the `ldap.user-member-attribute` attribute differently because this plugin aims first to load LDAP groups, secondly to load other useful attributes. You can think of the `{{ ldap.user-member-attribute }}=...` prefix am implicit

We covered this topic at length in our slack chat, please refer to that

mosiac1 commented:
I would like to get a consensus on `ldap.user-search-attributes` (or more generally on how do we want to handle fetching arbitrary user attributes from LDAP) before I go ahead and work on resolving the comments above. Once we agree on that the rest can be sorted out pretty quickly (famous last words).

cc: @Praveen2112 @dain

Praveen2112 commented:
I would like to get @dain's opinion here - on handling arbitrary user attributes.

dain commented:
I don't think we should the feature to map ldap attributes to groups.  Instead we should work on adding data from authentication to the session, and functions to access this data (so it can be used in ABAC expressions).  I'm not sure how we want to do this, but I'd start with a survey of how others have done this.  If there is a consensus from others, the design is simple.  This would be a separate project from this PR.

mosiac1 commented:
Thanks for the input Dain.

I agree we should have a better system for fetching and passing around user attributes (attributes = key-value string pairs), but I don't think it should be bound to the login process. It could it be its own separate plugin type (`UserAttributeProvider`?) that maps an `Identity` to a list of attributes. It could also be a new function in the `GroupProvider` interface since groups are really just a type of attribute.

For example, in my use-case, Trino is setup with both oauth and LDAP user/password authentication, but the user attributes I care about are only in LDAP.

I'm not seeing much support for the attribute to group mapping in this PR; that's ok, I will have to remove it. It may take a bit since this isn't a very high priority item for me currently.

SizintsevDA commented:
Thanks for the answer. Of course, we need to choose a more suitable solution. The main thing is that you can manage access rights using ldap groups right away without additional plugins. We tried the ldap group provider implementation, which is freely available. This is a good solution that is convenient to use.

ilsaloving commented:
Is there any progress on this?  I've been looking for a working group ldap plugin, and seeing years of endless arguing over non-core enhancements is somewhat frustrating.

I'm willing to help if I can, just to get this moving along.

mosiac1 commented:
We had chat about this PR during the last Trino contributor call https://github.com/trinodb/trino/wiki/Contributor-meetings#trino-contributor-call-24-oct-2024.

I don't have a lot of time to work on this unfortunately. We are looking for someone to take the PR to competition, its fairly close.

The discussion on the last few comments between me and Dain was about handling custom user attributes in LDAP. The decision was for that functionality to be removed. In my last push I stripped all of that out. The code for fetching groups from LDAP is in place.

The next steps are:
- Rebase
- Squash the last commit https://github.com/trinodb/trino/pull/20157/commits/154f76178ffac64fa0867bdff7511082ac6427ae into https://github.com/trinodb/trino/pull/20157/commits/d6f6539ce7ca5e42f66b650910604e30e24e65a7
- Go over all unresolved review comments and either address or resolve
  - A significant point here is removing the use of `@ParameterizedTests`
- Check and fix any relevant CI failures
- Get a review from Dain and/or David
- Write docs page

@OmerRaifler (I hope that is the right github user, slack user is https://trinodb.slack.com/team/U06BF92CABC) was interested in working on this, but I haven't heard anything from them

@ilsaloving Pleas feel free to jump in

mosiac1 commented:
This PR has now moved to https://github.com/trinodb/trino/pull/23900 as @OmerRaifler is taking ownership. We hope to get this shipped soon!
